x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskCleaner.java

```diff
@@ -0,0 +1,78 @@
+import org.elasticsearch.persistent.PersistentTasksCustomMetaData;
+import org.elasticsearch.threadpool.ThreadPool;
+
+/**
+ * A {@link ClusterStateListener} that completes any {@link ShardFollowTask} which concerns tao deleted index.
+ */
+public class ShardFollowTaskCleaner implements ClusterStateListener {
+
+    private static final Logger logger = LogManager.getLogger(ShardFollowTaskCleaner.class);
+
+    private final ThreadPool threadPool;
+    private final Client client;
+
+    public ShardFollowTaskCleaner(final ClusterService clusterService, final ThreadPool threadPool, final Client client) {
+        this.threadPool = threadPool;
+        this.client = client;
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public void clusterChanged(final ClusterChangedEvent event) {
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            return;
+        }
+        if (event.localNodeMaster()) {
+            final MetaData metaData = event.state().metaData();
+            final PersistentTasksCustomMetaData persistentTasksMetaData = metaData.custom(PersistentTasksCustomMetaData.TYPE);
+            if (persistentTasksMetaData != null) {
+                for (PersistentTasksCustomMetaData.PersistentTask<?> persistentTask : persistentTasksMetaData.tasks()) {
+                    if (ShardFollowTask.NAME.equals(persistentTask.getTaskName())) {
+                        ShardFollowTask shardFollowTask = (ShardFollowTask) persistentTask.getParams();
+                        Index followerIndex = shardFollowTask.getFollowShardId().getIndex();
+                        if (metaData.index(followerIndex) == null) {
+                            IndexNotFoundException e = new IndexNotFoundException(followerIndex);
+                            CompletionPersistentTaskAction.Request request =
+                                new CompletionPersistentTaskAction.Request(persistentTask.getId(), persistentTask.getAllocationId(), e);
+                            threadPool.generic().submit(() -> {
+                                client.execute(CompletionPersistentTaskAction.INSTANCE, request, new ActionListener<>() {
+
+                                    @Override
+                                    public void onResponse(PersistentTaskResponse persistentTaskResponse) {
+                                        logger.debug("task [{}] cleaned up", persistentTask.getId());
+                                    }
+
+                                    @Override
+                                    public void onFailure(Exception e) {
+                                        logger.warn("failed to clean up task [{}]", persistentTask.getId());
+                                    }
+                                });
+                            });
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
```

Comment:
The nesting here is pretty intense, how about rewriting some of the logic so that we don't get so deep. I think it helps make this clearer, too:

```diff
diff --git a/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskCleaner.java b/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskCleaner.java
index 758daa387d8..039a4695289 100644
--- a/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskCleaner.java
+++ b/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskCleaner.java
@@ -43,36 +43,45 @@ public class ShardFollowTaskCleaner implements ClusterStateListener {
         if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
             return;
         }
-        if (event.localNodeMaster()) {
-            final MetaData metaData = event.state().metaData();
-            final PersistentTasksCustomMetaData persistentTasksMetaData = metaData.custom(PersistentTasksCustomMetaData.TYPE);
-            if (persistentTasksMetaData != null) {
-                for (PersistentTasksCustomMetaData.PersistentTask<?> persistentTask : persistentTasksMetaData.tasks()) {
-                    if (ShardFollowTask.NAME.equals(persistentTask.getTaskName())) {
-                        ShardFollowTask shardFollowTask = (ShardFollowTask) persistentTask.getParams();
-                        Index followerIndex = shardFollowTask.getFollowShardId().getIndex();
-                        if (metaData.index(followerIndex) == null) {
-                            IndexNotFoundException e = new IndexNotFoundException(followerIndex);
-                            CompletionPersistentTaskAction.Request request =
-                                new CompletionPersistentTaskAction.Request(persistentTask.getId(), persistentTask.getAllocationId(), e);
-                            threadPool.generic().submit(() -> {
-                                client.execute(CompletionPersistentTaskAction.INSTANCE, request, new ActionListener<>() {
 
-                                    @Override
-                                    public void onResponse(PersistentTaskResponse persistentTaskResponse) {
-                                        logger.debug("task [{}] cleaned up", persistentTask.getId());
-                                    }
+        if (event.localNodeMaster() == false) {
+            return;
+        }
 
-                                    @Override
-                                    public void onFailure(Exception e) {
-                                        logger.warn("failed to clean up task [{}]", persistentTask.getId());
-                                    }
-                                });
-                            });
-                        }
-                    }
-                }
+        MetaData metaData = event.state().metaData();
+        PersistentTasksCustomMetaData persistentTasksMetaData = metaData.custom(PersistentTasksCustomMetaData.TYPE);
+        if (persistentTasksMetaData == null) {
+            return;
+        }
+
+        for (PersistentTasksCustomMetaData.PersistentTask<?> persistentTask : persistentTasksMetaData.tasks()) {
+            if (ShardFollowTask.NAME.equals(persistentTask.getTaskName()) == false) {
+                // this task is not a shard follow task
+                continue;
             }
+            ShardFollowTask shardFollowTask = (ShardFollowTask) persistentTask.getParams();
+            Index followerIndex = shardFollowTask.getFollowShardId().getIndex();
+            if (metaData.index(followerIndex) != null) {
+                // the index exists, do not clean this persistent task
+                continue;
+            }
+            IndexNotFoundException e = new IndexNotFoundException(followerIndex);
+            CompletionPersistentTaskAction.Request request =
+                new CompletionPersistentTaskAction.Request(persistentTask.getId(), persistentTask.getAllocationId(), e);
+            threadPool.generic().submit(() -> {
+                client.execute(CompletionPersistentTaskAction.INSTANCE, request, new ActionListener<>() {
+
+                    @Override
+                    public void onResponse(PersistentTaskResponse persistentTaskResponse) {
+                        logger.debug("task [{}] cleaned up", persistentTask.getId());
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.warn("failed to clean up task [{}]", persistentTask.getId());
+                    }
+                });
+            });
         }
     }
 }
```

Reply:
Sure. I borrowed your code in e1672b2c57875508c15c4bf9cf7839581da8fb52 if you don't mind.

Reply:
Of course not!